driver-core/src/main/com/mongodb/internal/Timeout.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal;
+
+import com.mongodb.annotations.Immutable;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.assertFalse;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * A <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/doc-files/ValueBased.html">value-based</a> class
+ * useful for tracking timeouts.
+ */
+@Immutable
+public final class Timeout {
+    private static final Timeout INFINITE = new Timeout(-1, 0);
```

Comment:
Could using `Long.MAX_VALUE` be a better / simpler constant for infinite?  That way you could remove the `isInfinite` assertion checks.

I know that in ReactiveStreams they use `Long.MAX_VALUE` to represent an infinite stream.

Reply:
We technically can use `Long.MAX_VALUE` to represent an infinite timeout internally (just replacing -1 with `Long.MAX_VALUE` on the line your highlighted is all that is required for such a change), but it will have no effect on the rest of the `Timeout` code, specifically it cannot result in removing any of the assertion checks. Which assertion check do you think would be gone as a result of such a change?

---
I would like to express some statements/considerations that I believe should partially explain the reasons behind the current API:

1. `startNow` interprets `Long#MAX_VALUE` as an infinite timeout (in addition to interpreting negative values the same way) because of the behaviour of `TimeUnit.convert` - it tolerates overflow by using `Long#MIN_VALUE`/`Long#MAX_VALUE` to represent a negative/positive overflow, thus making it impossible to differentiate between a situation when `Long#MAX_VALUE` nanoseconds is actually requested, and a situation when a longer duration is requested and an overflow happened.
2. The concept of remaining time is not applicable to an infinite timeout, so `remainingNanos` cannot be defined for it. `remainingNanosOrInfinite` is added to be used with APIs that interpret negative values as being infinite (e.g., `ConcurrentPool.get`). Having both `remainingNanos` that fails for infinite timeouts and `remainingNanosOrInfinite` helps informing a user (via either an error or a method name) that infinite timeouts cannot always be handled like other timeouts. These methods are similar to `Optional.get`/`Optional.orElseThrow` and `Optional.orElse` respectively, with the difference that `remainingNanosOrInfinite` does not allow specifying a value for an infinite timeout, while `Optional.orElse` allows specifying one for an empty `Optional.
3. `remainingNanosOrInfinite` uses a negative value instead of `Long.MAX_VALUE` for the following reasons:
3.1 [main reason] To further prevent a user from attempting to use infinite timeouts as any other timeouts by treating this value as an actual amount of remaining time. Even if I decided to return `Long.MAX_VALUE`, I would need to always return this constant, it could not go down as a result of multiple invocation of the `remainingNanosOrInfinite` method separated in time. Thus, all the internal `isInfinite` checks would still be required.
3.2 [supplementary reason] There are APIs accepting numeric timeout values and representing negative values as infinite timeouts (e.g., `ConcurrentPool.get`, maybe our code have more). I am not aware of APIs that do the same for `Long.MAX_VALUE` (even though this only makes sense given how `TimeUnit.convert` works).
4. Because `remainingNanosOrInfinite` must always return the same value for an infinite timeout,  there is no situation in which calling the method `remainingNanos(final long currentNanos)` would make sense for an infinite timeout. Therefore, it must not be called for such a timeout, which is why it has `assertFalse(isInfinite())`. Representing an infinite timeout with `Long.MAX_VALUE` internally would not have changed this.